### ISSUE TITLE
cli: support running cdk from inner directories

### ISSUE THREAD

**Issue by iliapolo** [aws/aws-cdk#5691]
I propose to have the `cdk` commands work even inside inner directories of the app itself, where there is no `cdk.json` file.

### Use Case

When working with `DockerImageAsset`, I think its rather common to have the following directory structure:

```text
hello-cdk/
├── cdk.json
├── app/
│   ├── Dockerfile
├── lib/
│   ├── hello-cdk-stack.ts
│   └── ...
└── bin/
```

Given this, I spend a lot of time inside the `hello-cdk/app` directory when working on the `Dockerfile`. On occasion I want to run `cdk deploy` to see the `cdk` is building my image properly. Problem is that `cdk` commands don't work in that directory since the `cdk.json` file is not there. This means having to `cd` out and back in quite a few times during the development cycle.

### Proposed Solution

My suggestion is that `cdk` will perform an up hill recursive search until in encounters a `cdk.json` file, similar to the way `git` commands work.

### Other

* [X] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

**Comment by eladb** [comment]
Sounds like a nice addition. Go for it

**Comment by iliapolo** [comment]
@eladb Well, I started having a look at this and ran into some complications.

Adding the traversal logic for loading the context and config files is easy, the problem though is this:

In `cdk.json` we have the `app` property:

```json
{
  "app": "npx ts-node bin/aws-cdk-playground.ts"
}
```

Its value is a command line that the cdk executes. Notice that it contains a relative path (`bin/aws-cdk-playground.ts`), which means it has to be executed from the project root directory.

There are a few possible solutions for this:

- Have the `cdk` _chdir_ to the project root (which is known because of the lookup) before running this command, and _chdir_ to the original one after.

I don't like this since there is no way to ensure the user won't end up in a different directory than the one he/she started in. This is very intrusive. 

- Parse this property and try to guess which part points to the file, and prepand the project root to that path. 

I don't like this since i'm not sure there is any heuristic we can apply to actually get this right for all use-cases. This command could be literaly anything.

- Restructure the config file to look something like:

```json
{
  "exec": "npx ts-node",
  "main": "bin/aws-cdk-playground.ts"
}
```

Given this config, we can simply prepand the project root to the value of the `main` property. 
But obvsiouly this would require some non trivial work to support backwards compatability.

Long story short, seems like implemeting this feature would require a greater effort, one which i think isn't worth it, given that no one has actually requrested this feature :)

Anywho, I'm dropping this for now and maybe we will get back to it later on, or, by popular demand if it gets enough likes.

**Comment by eladb** [comment]
> Have the cdk chdir to the project root (which is known because of the lookup) before running this command, and chdir to the original one after.
> I don't like this since there is no way to ensure the user won't end up in a different directory than the one he/she started in. This is very intrusive.

I think chdir is the correct behavior, otherwise we end up in a mess. A child process cannot change the working directory of a parent process, so there is no risk that the user will "end up in a different directory".

**Comment by iliapolo** [comment]
Hmm, good to know 👍 

I guess this happened to me only when `source`ing stuff...I'll take another stab at it.

**Comment by iliapolo** [comment]
Closing this issue manually since the PR didn't reference it at the time of merge.

**Comment by iliapolo** [comment]
I am re-opening this issue since the initial implenetation introduced a [bug](https://github.com/aws/aws-cdk/issues/5826), and we decided to revert it pending an implementation that doesn't cause it again.
